File: Python/BOJ/24479 알고리즘 수업 - 깊이 우선 탐색 1.py
import sys
cnt = 1
sys.setrecursionlimit(150000)


def dfs(t):
    global cnt
    visited[t] = cnt
    line[t].sort()
    for i in line[t]:
        if visited[i] == 0:
            cnt += 1
            dfs(i)


N, M, R = map(int, sys.stdin.readline().split())
line = [[] for _ in range(N+1)]
visited = [0]*(N+1)  # 저장값
for _ in range(M):
    a, b = map(int, sys.stdin.readline().split())
    line[a].append(b)  # 양 방향 간선
    line[b].append(a)  # 양 방향 간선
dfs(R)
for i in range(1, N+1):
    print(visited[i])

File: Java/BOJ/24479 알고리즘 수업 - 깊이 우선 탐색 1.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
	static int N;
	static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
	static int[] visited;
	static int cnt = 1;
	public static void dfs(int n) {
		visited[n] = cnt;
		Collections.sort(graph.get(n));
		for(int i = 0; i < graph.get(n).size(); i++) {
			int x = graph.get(n).get(i);
			if (visited[x] == 0) {
				cnt ++;
				dfs(x);
			}
		}
	}
	
	public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int R = Integer.parseInt(st.nextToken());
        for(int i = 0; i < N+1; i++) {
        	graph.add(new ArrayList<>());
        }
        visited = new int[N+1];
        for(int i = 0 ; i < M; i++) {
        	st = new StringTokenizer(br.readLine()," ");
        	int u = Integer.parseInt(st.nextToken());
        	int v = Integer.parseInt(st.nextToken());
        	graph.get(u).add(v);
        	graph.get(v).add(u);
        }

        dfs(R);
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i < N+1; i++) {
        	sb.append(visited[i]).append("\n");
        }
        System.out.println(sb);
    }
}